File: pkg/ui/workspaces/db-console/src/util/query.ts
// Copyright 2019 The Cockroach Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
// implied. See the License for the specific language governing
// permissions and limitations under the License.

import { Location } from "history";
import _ from "lodash";
import { match as Match } from "react-router-dom";

interface ParamsObj {
  [key: string]: string;
}

interface URLSearchParamsWithKeys extends URLSearchParams {
  keys: () => string;
}

// propsToQueryString is a helper function that converts a set of object
// properties to a query string
// - keys with null or undefined values will be skipped
// - non-string values will be toString'd
export function propsToQueryString(props: { [k: string]: any }): string {
  return _.compact(
    _.map(props, (v: any, k: string) =>
      !_.isNull(v) && !_.isUndefined(v)
        ? `${encodeURIComponent(k)}=${encodeURIComponent(v.toString())}`
        : null,
    ),
  ).join("&");
}

export function queryToObj(location: Location, key: string, value: string) {
  const params = new URLSearchParams(
    location.search,
  ) as URLSearchParamsWithKeys;
  const paramObj: ParamsObj = {};
  for (const data of params.keys()) {
    paramObj[data] = params.get(data);
  }
  if (key && value) {
    if (value.length > 0 || typeof value === "number") {
      paramObj[key] = value;
    } else {
      delete paramObj[key];
    }
  }
  return paramObj;
}

export function queryByName(location: Location, key: string): string | null {
  const urlParams = new URLSearchParams(location.search);
  return urlParams.get(key);
}

export function getMatchParamByName(match: Match<any>, key: string) {
  const param = match.params[key];
  if (param) {
    return decodeURIComponent(param);
  }
  return null;
}

File: pkg/sql/schemachanger/rel/query.go
// Copyright 2021 The Cockroach Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
// implied. See the License for the specific language governing
// permissions and limitations under the License.

package rel

import (
	"sort"

	"github.com/cockroachdb/errors"
	"github.com/semistrict/ratel/pkg/util"
	"github.com/semistrict/ratel/pkg/util/syncutil"
)

// Query searches for sets of entities which uphold a set of constraints.
type Query struct {
	schema *Schema
	// clauses are the original clauses. They exist for debugging.
	clauses []Clause
	// variables is the set of variables used in the query
	// stored in the order in which they appear.
	variables []Var
	// variableSlots is the mapping of names to slots.
	variableSlots map[Var]slotIdx
	// entities is the mapping of entities to slots.
	entities []slotIdx
	// slots store the data and metadata about the slots.
	slots []slot
	// facts are the set of facts which must be unified.
	facts []fact
	// filters are the set of predicate filters to evaluate.
	filters []filter

	// cache one evalContext for reuse to accelerate benchmarks and deal with
	// the common case.
	mu struct {
		syncutil.Mutex
		cached *evalContext
	}
}

// Result represents A setting of entities which fulfills the
// constraints of its corresponding query. It is a rather low-level
// interface.
type Result interface {

	// Var returns the value bound to the given variable.
	// If the variable does not exist in the query, nil will be
	// returned.
	Var(name Var) interface{}
}

// ResultIterator is used to iterate results of A query.
// Iteration can be halted with the use of iterutils.StopIteration.
type ResultIterator func(r Result) error

// NewQuery construct a new query with the provided clauses forming the
// conjunction of constraints on the results of the query when it is
// evaluated against a database.
func NewQuery(sc *Schema, clauses ...Clause) (_ *Query, err error) {
	defer func() {
		switch r := recover().(type) {
		case nil:
			return
		case error:
			err = errors.Wrap(r, "failed to construct query")
		default:
			err = errors.AssertionFailedf("failed to construct query: %v", r)
		}
	}()
	q := newQuery(sc, clauses)
	return q, nil
}

// Iterate will call the result iterator for every valid binding of each
// distinct entity variable such that all the variables in the query are
// bound and all filters passing.
func (q *Query) Iterate(db *Database, ri ResultIterator) error {
	ec := q.getEvalContext()
	defer q.putEvalContext(ec)
	return ec.Iterate(db, ri)
}

// getEvalContext grabs a cached evalContext from the query
// if one exists, otherwise it creates a new one.
func (q *Query) getEvalContext() *evalContext {
	getCachedEvalContext := func() (ec *evalContext) {
		q.mu.Lock()
		defer q.mu.Unlock()
		ec, q.mu.cached = q.mu.cached, ec
		return ec
	}
	if ec := getCachedEvalContext(); ec != nil {
		return ec
	}
	return newEvalContext(q)
}

// putEvalContext puts the evalContext in the cache if there is not one.
func (q *Query) putEvalContext(ec *evalContext) {
	q.mu.Lock()
	defer q.mu.Unlock()
	if q.mu.cached == nil {
		q.mu.cached = ec
	}
}

// Entities returns the entities in the query in their join order.
// This method exists primarily for introspection.
func (q *Query) Entities() []Var {
	var entitySlots util.FastIntSet
	for _, slotIdx := range q.entities {
		entitySlots.Add(int(slotIdx))
	}
	vars := make([]Var, 0, len(q.entities))
	for v, slotIdx := range q.variableSlots {
		if !entitySlots.Contains(int(slotIdx)) {
			continue
		}
		vars = append(vars, v)
	}
	sort.Slice(vars, func(i, j int) bool {
		return q.variableSlots[vars[i]] < q.variableSlots[vars[j]]
	})
	return vars
}

// Clauses returns the query's Clauses.
func (q *Query) Clauses() Clauses {
	return q.clauses
}
